hkmc2/shared/src/main/scala/hkmc2/semantics/ucs/DesugaringBase.scala

```diff
@@ -0,0 +1,122 @@
+
+  protected lazy val stringStartsWith: Ctxl[Term] =
+    term(Sel(Sel(Ident("globalThis"), Ident("Predef")), Ident("stringStartsWith")))
+
+  protected lazy val stringGet: Ctxl[Term] =
+    term(Sel(Sel(Ident("globalThis"), Ident("Predef")), Ident("stringGet")))
+
+  protected lazy val stringDrop: Ctxl[Term] =
+    term(Sel(Sel(Ident("globalThis"), Ident("Predef")), Ident("stringDrop")))
+
+  protected def callTupleGet(t: Term, i: Int, s: FlowSymbol): Ctxl[Term] =
+    app(tupleGet, tup(fld(t), fld(int(i))), s)
+
+  protected def callStringStartsWith(t: Term, prefix: Term, s: FlowSymbol): Ctxl[Term] =
+    app(stringStartsWith, tup(fld(t), fld(prefix)), s)
+
+  protected def callStringGet(t: Term, i: Int, s: FlowSymbol): Ctxl[Term] =
+    app(stringGet, tup(fld(t), fld(int(i))), s)
+
+  protected def callStringDrop(t: Term, n: Int, s: FlowSymbol): Ctxl[Term] =
+    app(stringDrop, tup(fld(t), fld(int(n))), s)
+
+  protected transparent inline def tempLet(term: Term)(inner: TempSymbol => Split): Split =
+    val s = TempSymbol(N, "temp")
+    Split.Let(s, term, inner(s))
+
+  protected transparent inline def plainTest(cond: Term, dbgName: Str = "cond")(inner: => Split): Split =
+    val s = TempSymbol(N, dbgName)
+    Split.Let(s, cond, Branch(s.ref(), inner) ~: Split.End)
+
+  /** Make a `Branch` that calls `Pattern` symbols' `unapply` functions. */
+  def makeUnapplyBranch(
+      scrut: => Term.Ref,
+      psym: PatternSymbol,
+      inner: => Split,
+      method: Str = "unapply"
+  )(fallback: Split): Ctxl[Split] =
+    val matchResultClassTerm = cls(matchResultClass)
+    matchResultClassTerm.symbol match
+      case S(matchResultClassSymbol: ClassSymbol) =>
+        // def makeUnapplyBranch(scrut: => Term.Ref, pat)
+        val resultIdent = Ident("matchResult"): Ident
+        val resultSymbol = TempSymbol(N, "matchResult")
+        val globalThis = term(Ident("globalThis"))
+        val unapply = Term.Sel(
+          prefix = Term.Sel(globalThis, Ident(psym.nme))(S(psym)),
+          nme = Ident(method)
+        )(N)
+        val arguments = Term.Tup(Fld(FldFlags.empty, scrut, N) :: Nil)(Tup(Nil))
+        val call = Term.App(unapply, arguments)(App(Empty(), Empty()), FlowSymbol(s"result of $method"))
+        Split.Let(resultSymbol, call,
+          Branch(
+            resultSymbol.ref(),
+            Pattern.ClassLike(matchResultClassSymbol, matchResultClassTerm, N, false)(matchResultClass),
+            inner
+          ) ~: fallback)
+      case S(_) | N => lastWords("Cannot locate `MatchResult` class in the global scope.")
+
+  /** Make a `Branch` that calls `Pattern` symbols' `unapplyStringPrefix` functions. */
+  def makeUnapplyStringPrefixBranch(
+      scrut: => Term.Ref,
+      psym: PatternSymbol,
+      inner: (scrut: TempSymbol) => Split,
+      method: Str = "unapplyStringPrefix"
+  )(fallback: Split): Ctxl[Split] =
+    val matchResultClassTerm = cls(matchResultClass)
```

Comment:
As I explained in the [other comment](https://github.com/hkust-taco/mlscript/pull/242#discussion_r1863721735), this is not the right approach. But also notice that you're recreating the term every time this function is called, making it doubly wrong.

Reply:
Thanks for the check. I've already modified this part of the code. Now I don't get symbols or construct terms through `cls` or `term`.